armi/utils/hexagon.py

```diff
@@ -191,3 +194,76 @@ def getIndexOfRotatedCell(initialCellIndex: int, orientationNumber: int) -> int:
     elif initialCellIndex == 1:
         return initialCellIndex
     raise ValueError(f"Cell number must be positive, got {initialCellIndex}")
+
+
+def rotateHexCellData(
+    data: typing.Union[list, np.ndarray], nRotations: int
+) -> typing.Union[list, np.ndarray]:
+    """Rotate data defined on a hexagonal lattice.
+
+    Consider a hexagon with three rings of cells. Each cell has a corresponding data, say a scalar
+    for its unique pin number. That hexagon is rotated 60 degrees counter clockwise. We need to update
+    these data so that the pin number for the pin originally in ring 2, position 1 travels
+    to where the pin ends up. The before and after pin numbers would look like::
+
+            9   8   7             7   18  17
+          10  2   1   18        8   1   6   16
+        11  3   0   6   17 -> 9   2   0   5   15
+          12  4   5   10        10  3   4   14
+            13  14  15            11  12  13
+
+    Parameters
+    ----------
+    data : list or numpy.ndarray
+        Data such that ``data[0]`` reflects data at the center of the hexagon,
+        ``data[1]`` is for ring 2, position 1, ``data[2]`` matches ring 2, position 2,
+        and generally ``data[i]`` corresponds to the ``i``-th cell in the ARMI spiraling
+        indexing scheme.
+    nRotations : int
+        Number of 60 degree counter clockwise rotations.
+
+    Returns
+    -------
+    list or numpy.ndarray
+        Data rotated accordingly. Matches the type of the initial ``data`` object so if a ``list``
+        was provided, a ``list`` will be returned.
+
+    Raises
+    ------
+    ValueError
+        If ``data`` does not have enough cells to exactly fill a hexagon with some number
+        of rings. For example, if your base hexagon has two rings (7 sites) but you do not
+        have data for the center location, passing an array with six entries is forbidden.
+        Place fake data like :attr:`math.nan` at that site.
+
+    """
+    if not isinstance(data, (list, np.ndarray)):
+        raise TypeError(f"{data=}")
+    cells = np.size(data, axis=0)
+    if cells == 0:
+        # Nothing to rotate
+        return data
+    nRings = numRingsToHoldNumCells(cells)
+    nCellsInFullRings = totalPositionsUpToRing(nRings)
+    if cells != nCellsInFullRings:
```

Comment:
Is there a way we can support this? We will have some blocks where the pins don't fill a full hex ring, but the rings do have an explicit position defined by a grid in the blueprints.

Would the onus be on the user/developer to cast this data into a full hexagon before trying to rotate, and then map it back after rotating? Maybe there is a way to automate that mapping.